src/geometry/partitions/distributedptscotchunstructuredpartition.h

```diff
@@ -0,0 +1,291 @@
+            int cy = from.x() + ((float)y / (float)resolution.y()) * dim.y();
+
+            if (region.count(Coord<2>(cx, cy)) > 0) {
+                os << '#';
+            } else {
+                os << '.';
+            }
+        }
+        os << '\n';
+    }
+}
+
+template<int DIM>
+class DistributedPTScotchUnstructuredPartition : public Partition<DIM>
+{
+public:
+    using Partition<DIM>::startOffsets;
+    using Partition<DIM>::weights;
+
+    DistributedPTScotchUnstructuredPartition(
+            const Coord<DIM> &origin,
+            const Coord<DIM> &dimensions,
+            const long &offset,
+            const std::vector<std::size_t> &weights,
+            const Adjacency &adjacency) :
+        Partition<DIM>(offset, weights),
+        origin(origin),
+        dimensions(dimensions),
+        numPartitions(weights.size())
+    {
+        this->adjacency = adjacency;
+
+        getStartEnd(MPILayer().rank(), start, end);
+        localCells = end - start;
+        buildRegions();
+    }
+
+    DistributedPTScotchUnstructuredPartition(
+            const Coord<DIM> &origin,
+            const Coord<DIM> &dimensions,
+            const long &offset,
+            const std::vector<std::size_t> &weights,
+            Adjacency &&adjacency) :
+        Partition<DIM>(offset, weights),
+        origin(origin),
+        dimensions(dimensions),
+        numPartitions(weights.size())
+    {
+        this->adjacency = std::move(adjacency);
+
+        getStartEnd(MPILayer().rank(), start, end);
+        localCells = end - start;
+        buildRegions();
+    }
+
+    void getStartEnd(int rank, size_t &start, size_t &end)
+    {
+        size_t cells = dimensions.prod();
+        start = (float)rank / numPartitions * cells;
+        end = (float)(rank + 1) / numPartitions * cells;
+    }
+
+    Region<DIM> getRegion(const std::size_t node) const override
+    {
+        return regions.at(node);
+    }
+
+private:
+    Coord<DIM> origin;
+    Coord<DIM> dimensions;
+    SCOTCH_Num localCells = 0;
+    size_t start = 0, end = 0;
+    size_t numPartitions = 0;
+    std::vector<Region<DIM> > regions;
+
+    void buildRegions()
+    {
+        std::vector<SCOTCH_Num> indices(localCells);
+
+        {
+            auto start = std::chrono::system_clock::now();
+            initIndices(indices);
+            auto duration = std::chrono::duration_cast<std::chrono::milliseconds> (std::chrono::system_clock::now() - start);
+            std::cout << "partitioning duration: " << duration.count() << std::endl;
+        }
+
+        regions.resize(numPartitions);
+
+        {
+            auto start = std::chrono::system_clock::now();
+            createRegions(indices);
+            auto duration = std::chrono::duration_cast<std::chrono::milliseconds> (std::chrono::system_clock::now() - start);
+            std::cout << "region creation duration: " << duration.count() << std::endl;
```

Comment:
For this use case we have the class ScopedTimer in src/misc/scopedtimer.h. See https://github.com/gentryx/libgeodecomp/blob/master/src/misc/test/unit/scopedtimertest.h for how to use it.